src/rgw/rgw_kmip_client.cc

```diff
@@ -0,0 +1,82 @@
+// -*- mode:C++; tab-width:8; c-basic-offset:2; indent-tabs-mode:t -*-
+// vim: ts=8 sw=2 smarttab ft=cpp
+
+#include "common/Thread.h"
+#include "include/compat.h"
+#include "common/errno.h"
+#include "rgw_common.h"
+#include "rgw_kmip_client.h"
+
+#include <atomic>
+
+#define dout_context g_ceph_context
+#define dout_subsys ceph_subsys_rgw
+
+RGWKMIPManager *rgw_kmip_manager;
+
+int
+RGWKMIPTransceiver::wait(optional_yield y)
+{
+	if (done)
+		return ret;
+	std::unique_lock l{lock};
+	if (!done)
+		cond.wait(l);
+	if (ret) {
+		lderr(cct) << "kmip process failed, " << ret << dendl;
+	}
+	return ret;
+}
+
+int
+RGWKMIPTransceiver::send()
+{
+	int r = rgw_kmip_manager->add_request(this);
+	if (r < 0) {
+		lderr(cct) << "kmip send failed, " << r << dendl;
+	}
+	return r;
+}
+
+int
+RGWKMIPTransceiver::process(optional_yield y)
+{
+	int r = send();
+	if (r < 0)
+		return r;
+	return wait(y);
+}
+
+RGWKMIPTransceiver::~RGWKMIPTransceiver()
+{
+	int i;
+	if (out)
+		free(out);
+	out = 0;
+	if (outlist->strings) {
+		for (i = 0; i < outlist->string_count; ++i) {
+			free(outlist->strings[i]);
+		}
+		free(outlist->strings);
```

Comment:
why is this on the heap? as opposed to, vector<string> or something?

Reply:
Ultimately it's a list of lists, so it's going to come off the heap no matter what.  This could all be coded using std::vector<std::string>;   That's probably not too costly, but there are other elements of RGWKMIPTransceiver where it gets more complicated.

So in src/rgw/rgw_kmip_client.h there's a declaration for RGWKMIPTransceiver.
outkey could be a string but that hides something I wish I could instead make more visible to the code: this isn't a utf8 string, it might contain embedded nulls, and it ought to be zeroized on free.  That's all true until it hits KmipGetTheKey::get_key_for_uniqueid where it gets turned into a std::string, at which point it's up tto the compiler if it actually does the "actual_key.replace" zeroize in src/rgw/rgw_crypt.cc.
out, name, unique_id are "char *"; there's logic that depends on being able to tell the difference between an unset value and a 0 length value which would need adjusting.

So doable, but complications.

Reply:
It's true, std::vector<std::string> uses the free store just as aggressively;  It does have the advantage of raii.  I'm not advocating you change this, honestly.  Just what was going through my mind reading the code.